### DISCUSSION TITLE
R: Roundtrip for arbitrary data types in ALTREP

### DISCUSSION THREAD

**Discussion by krlmlr** [duckdb/duckdb#8488] [Ideas]
`rel_from_df()` and `rel_to_altrep()` lose attributes of column vectors and also of the input data frame.

Is there a way to somehow capture additional column and table metadata so that it can be used during processing and materialization?

``` r
con <- DBI::dbConnect(duckdb::duckdb())

num <- structure(1:3, class = "foo", attr = "bar")

df <- tibble::tibble(num)

rel <- duckdb:::rel_from_df(con, df)
df_out <- duckdb:::rel_to_altrep(rel)

waldo::compare(df, df_out)
#> `class(old)`: "tbl_df" "tbl" "data.frame"
#> `class(new)`:                "data.frame"
#> 
#> `old$num` is an S3 object of class <foo>, an integer vector
#> `new$num` is an integer vector (1, 2, 3)
```

<sup>Created on 2023-08-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
I just realized that this can be implemented on the client.

**Comment by krlmlr** [comment]
Not quite, if we want the resulting data frame to look indistinguishable from the original. But we can solve it in the R glue, we don't necessarily need to go into duckdb proper for it.

@hannes @Mytherin: Is there a "proxy" vector/column type that is an indexed pointer into another vector? The main difference from a plain integer vector would be that no operations whatsoever are supported on that type. Happy to elaborate if this is unclear.

**Comment by krlmlr** [comment]
To rephrase: are there any user-exposed functions that operate on `LogicalType::POINTER` ?

**Comment by hannes** [comment]
I don't think there is a clean way of round tripping those through modestly complex queriess

**Reply by krlmlr** [reply]
I don't follow. Are you suggesting that `LogicalType::POINTER` can't be easily roundtripped, but `LogicalType::INTEGER` can? What's the difference if that column is never accessed in the query, but part of the input and present in the output?

Not sure I fully understand the semantics of `POINTER` .

**Reply by Tishj** [reply]
We use POINTER in the arrow scan function to carry function pointers that we cant otherwise hold on to

https://github.com/duckdb/duckdb/blob/acbbfe0e79fdfef712a8dc317af407c12006814f/tools/pythonpkg/src/python_udf.cpp#L61

https://github.com/duckdb/duckdb/blob/acbbfe0e79fdfef712a8dc317af407c12006814f/src/function/table/arrow.cpp#L213

EDIT:
btw I definitely wasn't trying to bud in on the question at hand, just providing some context on how POINTER is used in other places currently

**Reply by hannes** [reply]
Yes we could move pointers through the engine, that's not the issue, but how to do that for per-column R attributes is not yet clear to me. Also would mean query generation has to be very careful to drag those along correctly.

**Reply by krlmlr** [reply]
My plan:

- when ingesting a data frame, decide for every column if it can be mapped losslessly to a duckdb data type
- columns that can't be mapped are replaced by `POINTER` columns. These pointer columns will have to refer to the original data frame in some way, not yet sure how. As a last resort, we could use a `STRUCT` of two `POINTER`s
- if a non-mapped column is accessed in a query, we bail out and fall back to the existing implementation
- when constructing the ALTREP data frames, `POINTER` columns are replaced with the result of applying `df[idx, ]` in R land, obeying whatever rules the implementation mandates.

Happy to offer a POC implementation.

**Comment by krlmlr** [comment]
- Custom column types: see experimental R string implementation, `LogicalType RStringsType::Get()` .
- Logical type that doesn't consume memory: `SQLNULL` .

**Reply by Tishj** [reply]
> - Custom column types: see experimental R string implementation, `LogicalType RStringsType::Get()` .
> - Logical type that doesn't consume memory: `SQLNULL` .

Ah okay, I see.
What is the reason we can't use a custom ExtraTypeInfo instead, what extra functionality does this provide?

**Reply by krlmlr** [reply]
I'll look into custom `ExtraTypeInfo` too.
